lib/lualink.c

```diff
@@ -537,28 +543,71 @@ void Lua_crowbegin( void )
 {
     printf("init()\n"); // call in C to avoid user seeing in lua
     lua_getglobal(L,"init");
-    if( lua_pcall_protected(L,0,0,0) != LUA_OK ){
-        Caw_send_luachunk( (char*)lua_tostring( L, -1 ) );
-        lua_pop( L, 1 );
+    if( Lua_call_usercode(L,0,0) != LUA_OK ){
+        Caw_send_luachunk("error running init()");
+        lua_pop(L, 1);
     }
 }
 
 
 // Watchdog timer for infinite looped Lua scripts
 volatile int watchdog = WATCHDOG_COUNT;
-static void hook( lua_State* L, lua_Debug* ar )
+static void timeouthook( lua_State* L, lua_Debug* ar )
 {
     if( --watchdog <= 0 ){
         Caw_send_luachunk("CPU timed out.");
-        lua_sethook(L, hook, LUA_MASKLINE, 0); // error until top
-        luaL_error(L,"");
+        lua_sethook(L, timeouthook, LUA_MASKLINE, 0); // error until top
+        luaL_error(L, "user code timeout exceeded");
+    }
+}
+
+// Check for too much recursion
+static void callhook( lua_State* L, lua_Debug* ar )
+{
+    for (int level = 0; lua_getstack(L, level, ar); level++) {
+        if (level > MAX_CALLFRAMES) {
+	    Caw_send_luachunk("Call depth exceeded " TOSTRING(MAX_CALLFRAMES));
+	    luaL_error(L, "recursion limit exceeded");
+	}
     }
 }
-static int lua_pcall_protected( lua_State* L, int nargs, int nresults, int errFunc )
+
+static int Lua_handle_error( lua_State *L )
+{
+    const char *msg = lua_tostring( L, 1 );
+    if( msg == NULL ){
+        if( luaL_callmeta( L, 1, "__tostring" )
+         && lua_type ( L, -1 ) == LUA_TSTRING ) {
+            return 1;
+        } else {
+            msg = lua_pushfstring( L
+                                 , "(error object is a %s value)"
+                                 , luaL_typename( L, 1 ) );
+        }
+    }
+    luaL_traceback( L, L, msg, 1 );
+    char* traceback = (char*)lua_tostring( L, -1 );
+    Caw_send_luachunk( traceback );
+    _printf( traceback );
+    return 1;
+}
+
+static int Lua_call_usercode( lua_State* L, int nargs, int nresults )
 {
-    lua_sethook(L, hook, LUA_MASKCOUNT, WATCHDOG_FREQ); // reset hook
-    watchdog = WATCHDOG_COUNT; // reset hook counter
-    return lua_pcall(L, nargs, nresults, errFunc);
+    lua_sethook(L, timeouthook, LUA_MASKCOUNT, WATCHDOG_FREQ); // reset timeout hook
+    watchdog = WATCHDOG_COUNT; // reset timeout hook counter
+    lua_sethook(L, callhook, LUA_MASKCALL, 0); // reset call hook
+
+    int errFunc = lua_gettop(L) - nargs;
+    lua_pushcfunction( L, Lua_handle_error );
+    lua_insert( L, errFunc );
+    int status = lua_pcall(L, nargs, nresults, errFunc);
+    lua_remove( L, errFunc );
+
+    lua_sethook(L, timeouthook, 0, 0);
+    lua_sethook(L, callhook, 0, 0);
```

Comment:
Just curious why these need to be unset after ending the code execution? Is it just a sanitation thing?

Reply:
Yeah just habitually cleaning up the state, don't think it matters, though it would for the `callhook` if we only want to install it for calling `init()`